### DISCUSSION TITLE
Module/symbol rename with direct write to disk (no open editors) and other suggestions

### DISCUSSION THREAD

**Discussion by smackesey** [microsoft/pylance-release#3083] [Enhancement]
Module/symbol rename is most useful on large projects. I work on a large project ([Dagster](https://github.com/dagster-io/dagster)) where we needed to rename core submodules to mark them private. I just used Pylance to rename a module in VSCode, though it got the job done, the user experience was not great. Here were the issues:

- The rename operation took a significant amount of time, and there was no indicator of progress. There was an endlessly cycling progress bar in the upper left. After a while when my CPU had calmed down, the cycling was still happening but I didn't know if processing was done or not-- the preview list seemed to be complete. I decided to just hit Apply and it worked, but it was confusing since the "progress bar" was still cycling.
- Mysteriously during the generation of the rename preview, the "Problem" counter started to rise a little.
- After clicking "Apply", VSCode opened 951 editors for the files with changed imports and apparently analyzed all of them for diagnostics, resulting in a "Problems" counter of 11K+.
- After clicking "Save All', VSCode successfully saved all the files, but then my 20-core Mac Studio completely locked up with spinning beach ball. I waited for like 10 minutes hoping it would end gracefully, but I eventually had to use the physical power switch to reboot.

Fortunately I'm on a fast machine, so all steps until the last weren't too painful. But I imagine it would have been much worse on an older laptop. IMO this is how it should work:

- VSCode shows you a progress bar with the number of files it is searching for references. The progress bar completes when all target files have been searched. There is no cycling progress indicator once the search is complete.
- Clicking "Apply" presents a popup where you are given the option of writing changes directly to disk _or_ applying them in-editor (the current behavior).
- There is some kind of configurable resource limit on Pylance (with good defaults) so that it doesn't crash your machine in an effort to deal with 950 updated files.

**Comment by judej** [comment]
Moving this issue to discussion as an enhancement request for comments and upvotes.
